staging/kos/pkg/controllers/subnet/validator.go

```diff
@@ -0,0 +1,433 @@
+		v.queue.Add(k8stypes.NamespacedName{
+			Namespace: newS.Namespace,
+			Name:      newS.Name,
+		})
+	}
+}
+
+func (v *Validator) OnDelete(obj interface{}) {
+	s := parse.Peel(obj).(*netv1a1.Subnet)
+	glog.V(5).Infof("Notified of deletion of %#+v.", s)
+	v.queue.Add(k8stypes.NamespacedName{
+		Namespace: s.Namespace,
+		Name:      s.Name,
+	})
+}
+
+func (v *Validator) processQueue() {
+	for {
+		subnet, stop := v.queue.Get()
+		if stop {
+			return
+		}
+		v.processQueueItem(subnet.(k8stypes.NamespacedName))
+	}
+}
+
+func (v *Validator) processQueueItem(subnet k8stypes.NamespacedName) {
+	defer v.queue.Done(subnet)
+	requeues := v.queue.NumRequeues(subnet)
+	if err := v.processSubnet(subnet); err != nil {
+		glog.Warningf("Failed processing %s, requeuing (%d earlier requeues): %s.", subnet, requeues, err.Error())
+		v.queue.AddRateLimited(subnet)
+		return
+	}
+	glog.V(4).Infof("Finished %s with %d requeues.", subnet, requeues)
+	v.queue.Forget(subnet)
+}
+
+func (v *Validator) processSubnet(subnetNSN k8stypes.NamespacedName) error {
+	subnet, err := v.subnetLister.Subnets(subnetNSN.Namespace).Get(subnetNSN.Name)
+
+	if err != nil && !k8serrors.IsNotFound(err) {
+		glog.Errorf("subnet lister failed to lookup %s: %s", subnetNSN, err.Error())
+		// This should never happen. No point in retrying.
+		return nil
+	}
+
+	if k8serrors.IsNotFound(err) {
+		v.processDeletedSubnet(subnetNSN)
+		return nil
+	}
+
+	return v.processExistingSubnet(subnet)
+}
+
+func (v *Validator) processDeletedSubnet(s k8stypes.NamespacedName) {
+	v.clearStaleRV(s)
+
+	rivals := v.clearConflictsCache(s)
+
+	// Enqueue old rivals so that they can be re-validated: they might no longer
+	// have conflicts as this subnet has been deleted.
+	for _, r := range rivals {
+		v.queue.Add(r)
+	}
+}
+
+func (v *Validator) processExistingSubnet(s *netv1a1.Subnet) error {
+	ss, parsingErrs := subnet.NewSummary(s)
+	if len(parsingErrs) > 0 {
+		return parsingErrs
+	}
+
+	if v.subnetIsStale(ss.NamespacedName, s.ResourceVersion) {
+		return nil
+	}
+
+	// If we're here s might have been created or updated in a way that affects
+	// validation. We need to update its conflicts cache accordingly and
+	// reconsider old rivals because they might no longer be in conflict with
+	// s.
+	oldRivals := v.updateConflictsCache(ss)
+	for _, r := range oldRivals {
+		v.queue.Add(r)
+	}
+
+	// Keep the promise that a Subnet stays validated once it becomes validated
+	// (unless and until its VNI or CIDR block changes).
+	if s.Status.Validated {
+		return nil
+	}
+
+	// Retrieve all the other subnets with the same VNI. Doing a live list as
+	// opposed to a cache-based one (through the informer) prevents race
+	// conditions that can arise in case of multiple validators running.
+	potentialRivals, err := v.netIfc.Subnets(k8scorev1api.NamespaceAll).List(k8smetav1.ListOptions{
+		FieldSelector: k8sfields.OneTermEqualSelector(subnetVNIField, fmt.Sprint(ss.VNI)).String(),
```

Comment:
Before the latest commits I realized that both were being used, so I switched to using `fmt.Sprint` everywhere for consistency. I chose it over the alternative because it's more concise. Do you have an argument for using `strconv.FormatUint`? It's certainly more explicit than `fmt.Sprint`, but it's more verbose. I see no clear winner.

Reply:
On further reflection, you might be right. `fmt.Sprint` and `strconv.FormatUint` happen to do the same thing, but in `strconv.FormatUint` the names guarantee that the result is what we want, in `fmt.Sprint` we have no such guarantee. Also, package `fmt` uses reflection, it's likely that it is slower.

Reply:
Yes, I tend to avoid reflection when it is unnecessary.